README.md
# Starfruit

A library containing useful utilities, efficiencies, and 
abstractions to improve over the Dart SDK and dart:math. It takes heavy 
inspiration from Guava (by Google) for Java, and ports some functions from 
it. 

---

It takes time, effort, and mental power to keep this package updated, useful, and
improving. If you used or are using the package, I'd appreciate it if you could spare a few 
dollars to help me continue to do so.

[![PayPal](https://img.shields.io/static/v1?label=PayPal&message=Donate&color=blue&logo=paypal&style=for-the-badge&labelColor=black)](https://www.paypal.me/kishoredev)

---

## Classes

Note: https://pub.dev/documentation/starfruit/latest/starfruit/starfruit-library.html 
is auto-generated dartdocs. This has a more detailed rundown of every class
and method. It's impossible to list every method offered and its purpose, 
so I won't try. You **should** use the usage examples and documentation to understand
the utilities of this library. For here, I'm just offering basic details on how to 
instantiate every class.

##### StarMathUtils (mathUtils)
To use, call methods on the singleton instance mathUtils (e.g. `mathUtils.chunks([1, 2, 3, 4, 5, 6], 2)` ).

##### StarCollectionUtils (collectionUtils)
To use, call methods on the singleton instance collectionUtils (e.g. `collectionUtils.chunks([1, 2, 3, 4, 5, 6], 2)` ).

##### StarStats (`List<num>`)
To use, collect a list of numerical input data. You can then call StarStats methods on it, because
StarStats is defined as an extension on List<num>.

##### StarStatsXY (```Map<num,num>``` of points)
To use, collect input data formatted {x1: y1, x2: y2, x3: y3, ...}. You can then call StarStatsXY methods
on it, because StarStatsXY is defined as an extension on Map<num, num>.

---

## Usage Examples

##### Math Utilities:

```dart
import 'package:starfruit/starfruit.dart';

main() {
  /Round to 2 decimal places
  //Ceiling and flooring to decimal places is also available
  print('Round 3.5634 to 2 decimal places:');
  print(mathUtils.roundToDouble(3.5634, 2));
  print('');

  //Power of two check
  print('Is 8 a power of two?');
  print(mathUtils.isPowerOfTwo(8));
  print('');

  //Calculate log base 2 of 8
  print('Calculate log base 2 of 8:');
  print(mathUtils.log(2, 8));
  print('');

  //double isInteger check
  print('Is 6.0 an integer?');
  print(mathUtils.isMathematicalInteger(6.0));
  print('');

  //Calculate 5!
  print('Calculate 5 factorial:');
  print(mathUtils.factorial(5));
  print('');

  //Check if 2 is within 3 of 5
  print('Is 2 within 3 of 5?');
  print(mathUtils.fuzzyEquals(2, 5, 3));
  print('');

  //Check if 7919 is a prime number
  print('Is 7919 prime?');
  print(mathUtils.isPrime(7919));
  print('');

  //LCM of 2 and 7
  print('Calculate LCM of 2 and 7:');
  print(mathUtils.lcm(2, 7));
  print('');

  //sinh of 1 radian
  print('Calculate sinh of 1 radian:');
  print(mathUtils.sinh(1));
  print('');

  //mean of 3 and 7
  print('Calculate mean of 3 and 7:');
  print(mathUtils.mean(3,7));
  print('');

  //nCr where n = 5 and r = 2
  print('Calculate the binomial coefficient of 2 and 5:');
  print(mathUtils.combinationsOf(5, 2));
  print('');

  //nPr where n = 5 and r = 2
  print('Calculate the permutations of 5 and 2:');
  print(mathUtils.permutationsOf(5, 2));
  print('');
}
```
##### Collection Utilities:
```dart
import 'package:starfruit/starfruit.dart';

main() {
  //Separate List into chunks
  print("Separate List into chunks of 4:");
  print(cUtils.chunks([1,2,3,4,5,6,7,8], 4));
  print("");

  //Collapse List<List<Object>> into just List<Object>
  print("Collapse List<List> into List:");
  print(cUtils.collapse( [[1,2,3],[4,5,6,7,8]] ));
  print("");

  //Get every nth object of list
  print("Get every 4th object:");
  print(cUtils.nth([1,2,3,4,5,6,7,8], 4));
  print("");

  //Get every n random elements of list
  print("Get 5 random elements:");
  print(cUtils.random([1,2,3,4,5,6,7,8], 5));
  print("");

  //Flip keys and values of a map
  print("Flip keys and values of a map:");
  print(cUtils.flip({1:0, 2:8, 3:4, 0:5}));
  print("");

  //Zip two Lists into a map
  print("Zip two lists into a map:");
  print(cUtils.zip([1,2,3,4], [5,6,7,8]));
  print("");

  //Unzip a map into two lists
  print("Unzip a map into two lists:");
  print(cUtils.unzip({0:1, 8:2, 4:3, 5:0}));
  print("");
}
```

##### Stats Utilities:

```dart
import 'package:starfruit/starfruit.dart';

main() {
  var stats = [1,3,4,5,12,3,4,67,8,0,22];

  //median
  print("Calculate the median:");
  print(stats.median);
  print("");

  //mean
  print("Calculate the mean:");
  print(stats.mean);
  print("");

  //mode
  print("Calculate the mode:");
  print(stats.mode);
  print("");

  //variance
  print('Calculate the variance:');
  print(stats.variance);
  print("");

  //standard deviation
  print("Calculate the standard deviation:");
  print(stats.stdDev);
  print("");

  //cardinality; number of unique elements
  print("Calculate the cardinality:");
  print(stats.cardinality);
  print("");

  //7 greatest elements
  print("Return the 7 greatest elements:");
  print(stats.topElements(7));
  print("");
}
```
##### XY (Paired) Stats Utilities:

```dart
import 'package:starfruit/starfruit.dart';

main() {
  var xystats = {2:3, 7:4, 4:6, 8:9, 1:2, 3:5, 11:14, 12:18};

  //Get correlation coefficient
  print("Calculate correlation coefficient:");
  print(xystats.corCoefficient);
  print("");

  //Get determination coefficent
  print("Calculate determination coefficient:");
  print(xystats.detCoefficient);
  print("");

  //Get adjusted determination coefficient
  print("Calculate adj. determination coefficient:");
  print(xystats.adjDetCoefficient);
  print("");

  //Get linear regression line
  print("Calculate linear regression line in form y = mx + b:");
  var lr = xystats.linReg;
  print("y = ${lr[0]}x + ${lr[1]}");
  print("");

  //Get quadratic regression line
  print("Calculate quadratic regression line in form y = ax^2 + bx + c:");
  var qdr = xystats.quadReg;
  print("y = ${qdr[0]}x^2 + ${qdr[1]}x + ${qdr[2]}");
  print("");

  //Get cubic regression line
  print("Calculate cubic regression line in form y = ax^3 + bx^2 + cx + d:");
  var cr = xystats.cubicReg;
  print("y = ${cr[0]}x^3 + ${cr[1]}x^2 + ${cr[2]}x + ${cr[3]}");
  print("");

  //Get quartic regression line
  print("Calculate quartic regression line in form y = ax^4 + bx^3 + cx^2 + dx + e:");
  var qrr = xystats.quarReg;
  print("y = ${qrr[0]}x^4 + ${qrr[1]}x^3 + ${qrr[2]}x^2 + ${qrr[3]}x + ${qrr[4]}");
  print("");
}
```

---

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/AKushWarrior/starfruit/issues

---

[![Pub](https://img.shields.io/pub/v/starfruit?color=blue&label=pub&logo=Steel%20Crypt&logoColor=blue&style=for-the-badge&labelColor=black)](https://pub.dev/packages/starfruit)
[![License](https://img.shields.io/github/license/AKushWarrior/starfruit?color=blue&style=for-the-badge&labelColor=black)](https://opensource.org/licenses/lgpl-3.0.html)
[![Commits](https://img.shields.io/github/commit-activity/m/AKushWarrior/starfruit?color=blue&style=for-the-badge&labelColor=black)](https://github.com/AKushWarrior/starfruit)


###### Starfruit, a set of Dart utility libraries.
###### ©2020 Aditya Kishore
###### Licensed under the MPL 2.0 (See LICENSE)


CHANGELOG.md
## 2.0.0
- Null Safety
- Fixed minor bugs
- No functional change from 1.0.0
- Add population-based sample and variance methods.

## 1.0.0 - 1.0.0+1
- Fixed bugs
- Reworked APIs to extensions where possible
- Started large test suite
- Relicensed to MPL 2.0

## 0.3.1+1
- Major README issue
- No functional change

## 0.3.1
- ```StarLinkedList()``` introduced
- Documentation edits
- Significant README revamp

## 0.2.3+1 (9/2/2019)
- Hotfix for misrepresentative README issue

## 0.2.3 (9/2/2019)
- Introducing Cubic and Quartic Regression to ```StarStatsXY()```
- README and documentation edits

## 0.2.2 (9/2/2019)
- Introducing Quadratic Regression to ```StarStatsXY()```
- README and documentation edits

## 0.2.1 (8/30/2019)
- Introducing ```StarCollectionUtils()```
- Documentation improvement
- README edits

## 0.1.1 (8/30/2019)
- Initial version
- Built and finished building ```StarMathUtils()```
- Started ```StarStats()``` and ```StarStatsXY()```
- More to come...


test/mathUtilTest.dart
import 'package:test/test.dart';
import 'package:should/should.dart';
import 'package:starfruit/starfruit.dart';

void main() {
  should.errorOnAssert = true;

  group('roundToDouble should round to the specified number of places:', () {
    test('two places rounding up', () {
      var rounded = mathUtils.roundToDouble(1.005, 2);
      requireThat(rounded).equal(1.01);
    });
    test('two places rounding down', () {
      var rounded = mathUtils.roundToDouble(1.013, 2);
      requireThat(rounded).equal(1.01);
    });
    test('negative two places rounding down', () {
      var rounded = mathUtils.roundToDouble(-1.014, 2);
      requireThat(rounded).equal(-1.01);
    });
  });

  test('ceilToDouble should ceil to the specified number of places', () {
    var rounded = mathUtils.ceilToDouble(1.002, 2);
    requireThat(rounded).equal(1.01);
  });

  test('floorToDouble should floor to the specified number of places', () {
    var rounded = mathUtils.floorToDouble(1.017, 2);
    requireThat(rounded).equal(1.01);
  });

  group('isPowerOfTwo should return true for all powers of two:', () {
    test('non-power of two (34)', () {
      requireThat(mathUtils.isPowerOfTwo(34)).equal(false);
    });
    test('non-power of two (-2)', () {
      requireThat(mathUtils.isPowerOfTwo(-2)).equal(false);
    });
    test('power of two (8)', () {
      requireThat(mathUtils.isPowerOfTwo(8)).equal(true);
    });
    test('power of two (0.25)', () {
      requireThat(mathUtils.isPowerOfTwo(0.25)).equal(true);
    });
  });

  group('isPowerOfN should return true for all powers of N:', () {
    test('non-power of three (34, 3)', () {
      requireThat(mathUtils.isPowerOfN(34, 3)).equal(false);
    });
    test('power of four (16, 4)', () {
      requireThat(mathUtils.isPowerOfN(16, 4)).equal(true);
    });
    test('power of negative 2 (-2, -8)', () {
      requireThat(mathUtils.isPowerOfN(-8, -2)).equal(true);
    });
    test('power of negative 3 (-3, 9)', () {
      requireThat(mathUtils.isPowerOfN(9, -3)).equal(true);
    });
  });

  test('logBase should return what exponent of the base the input is', () {
    requireThat(mathUtils.logBase(2, 8)).equal(3);
  });

  test('isMathematicalInteger should return true if input is divisible by 1',
      () {
    requireThat(mathUtils.isMathematicalInteger(3.0)).equal(true);
  });

  group('factorial should return the factorial of the input:', () {
    test('factorial 5', () {
      var fac5 = mathUtils.factorial(5);
      requireThat(fac5).equal(120);
    });

    test('factorial 13', () {
      var fac13 = mathUtils.factorial(13);
      requireThat(fac13).equal(6227020800);
    });
  });

  group('fuzzy equals should return whether the two numbers are within a range of each other:', () {
    test('6 and 4, range 2', () {
      var base = mathUtils.fuzzyEquals(6, 4, 2);
      requireThat(base).equal(true);
    });

    test('4 and 6, range 2', () {
      var base = mathUtils.fuzzyEquals(4, 6, 2);
      requireThat(base).equal(true);
    });

    test('7 and 11, range 3', () {
      var base = mathUtils.fuzzyEquals(7, 11, 3);
      requireThat(base).equal(false);
    });

    test('11 and 7, range 3', () {
      var base = mathUtils.fuzzyEquals(11, 7, 3);
      requireThat(base).equal(false);
    });
  });

  group('isPrime checks if number is a prime:', () {
    test('1', () {
      requireThat(mathUtils.isPrime(1)).equal(false);
    });
    test('0', () {
      requireThat(mathUtils.isPrime(0)).equal(false);
    });
    test('7907', () {
      requireThat(mathUtils.isPrime(7907)).equal(true);
    });
    test('2', () {
      requireThat(mathUtils.isPrime(2)).equal(true);
    });
  });
}


test/math_util_test.dart
import 'package:test/test.dart';
import 'package:starfruit/starfruit.dart';

void main() {
  group('roundToDouble should round to the specified number of places:', () {
    test('two places rounding up', () {
      var rounded = mathUtils.roundToDouble(1.005, 2);
      expect(rounded, equals(1.01));
    });
    test('two places rounding down', () {
      var rounded = mathUtils.roundToDouble(1.013, 2);
      expect(rounded, equals(1.01));
    });
    test('negative two places rounding down', () {
      var rounded = mathUtils.roundToDouble(-1.014, 2);
      expect(rounded, equals(-1.01));
    });
  });

  test('ceilToDouble should ceil to the specified number of places', () {
    var rounded = mathUtils.ceilToDouble(1.002, 2);
    expect(rounded, equals(1.01));
  });

  test('floorToDouble should floor to the specified number of places', () {
    var rounded = mathUtils.floorToDouble(1.017, 2);
    expect(rounded, equals(1.01));
  });

  group('isPowerOfTwo should return true for all powers of two:', () {
    test('non-power of two (34)', () {
      expect(mathUtils.isPowerOfTwo(34), equals(false));
    });
    test('non-power of two (-2)', () {
      expect(mathUtils.isPowerOfTwo(-2), equals(false));
    });
    test('power of two (8)', () {
      expect(mathUtils.isPowerOfTwo(8), equals(true));
    });
    test('power of two (0.25)', () {
      expect(mathUtils.isPowerOfTwo(0.25), equals(true));
    });
  });

  group('isPowerOfN should return true for all powers of N:', () {
    test('non-power of three (34, 3)', () {
      expect(mathUtils.isPowerOfN(34, 3), equals(false));
    });
    test('power of four (16, 4)', () {
      expect(mathUtils.isPowerOfN(16, 4), equals(true));
    });
    test('power of negative 2 (-8, -2)', () {
      expect(mathUtils.isPowerOfN(-8, -2), equals(true));
    });
    test('power of negative 3 (9, -3)', () {
      expect(mathUtils.isPowerOfN(9, -3), equals(true));
    });
  });

  test('logBase should return what exponent of the base the input is', () {
    expect(mathUtils.logBase(2, 8), equals(3));
  });

  test('isMathematicalInteger should return true if input is divisible by 1',
      () {
    expect(mathUtils.isMathematicalInteger(3.0), equals(true));
  });

  group('factorial should return the factorial of the input:', () {
    test('factorial 5', () {
      var fac5 = mathUtils.factorial(5);
      expect(fac5, equals(120));
    });

    test('factorial 13', () {
      var fac13 = mathUtils.factorial(13);
      expect(fac13, equals(6227020800));
    });
  });

  group('fuzzy equals should return whether the two numbers are within a range of each other:', () {
    test('6 and 4, range 2', () {
      var base = mathUtils.fuzzyEquals(6, 4, 2);
      expect(base, equals(true));
    });

    test('4 and 6, range 2', () {
      var base = mathUtils.fuzzyEquals(4, 6, 2);
      expect(base, equals(true));
    });

    test('7 and 11, range 3', () {
      var base = mathUtils.fuzzyEquals(7, 11, 3);
      expect(base, equals(false));
    });

    test('11 and 7, range 3', () {
      var base = mathUtils.fuzzyEquals(11, 7, 3);
      expect(base, equals(false));
    });
  });

  group('isPrime checks if number is a prime:', () {
    test('1', () {
      expect(mathUtils.isPrime(1), equals(false));
    });
    test('0', () {
      expect(mathUtils.isPrime(0), equals(false));
    });
    test('7907', () {
      expect(mathUtils.isPrime(7907), equals(true));
    });
    test('2', () {
      expect(mathUtils.isPrime(2), equals(true));
    });
  });
}


lib/src/math/mathutils.dart
/*
Starfruit, a set of Dart utility libraries.
Copyright (C) 2020 Aditya Kishore

This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
*/

import 'dart:math';
import 'dart:math' as math;

///Default implementation of StarMathUtils().
_StarMathUtils get mathUtils => _StarMathUtils();

///A class of packaged math utilities and various commonplace functions not
///provided by (or improving upon) Dart functions. They can be used the default
///construction of StarMathUtils, mUtils.
class _StarMathUtils {
  ///Round to a double place where the second input corresponds to the amount
  ///of decimal places in the first double (eg 1 ==> x.x, 2 ==> x.xx, 3 ==> x.xxx).
  ///
  /// Note that, due to floating point errors, this method doesn't work past
  /// ~14 decimal places.
  num roundToDouble(num input, int decPlaces) {
    for (int i = 0; i < decPlaces; i++) {
      input *= 10;
    }
    input += 0.00000000000001;
    input = input.round();
    return input / pow(10, decPlaces);
  }

  ///Round up (ceil) to a double place where the second input corresponds to the amount
  ///of decimal places in the first double (eg 1 ==> x.x, 2 ==> x.xx, 3 ==> x.xxx).
  double ceilToDouble(num input, int decPlaces) {
    var fac = pow(10, decPlaces);
    input = (input * fac).ceil() / fac;
    return input.toDouble();
  }

  ///Round down (floor) to a double place where the second input corresponds to the amount
  ///of decimal places in the first double (eg 1 ==> x.x, 2 ==> x.xx, 3 ==> x.xxx).
  double floorToDouble(num input, int decPlaces) {
    var fac = pow(10, decPlaces);
    input = (input * fac).floor() / fac;
    return input.toDouble();
  }

  ///Returns whether given ```input``` is a perfect power of two.
  bool isPowerOfTwo(num input) {
    if (input <= 0) {
      return false;
    }
    return isMathematicalInteger(logBase(2, input));
  }

  ///Returns whether given ```input``` is a perfect power of `n`.
  bool isPowerOfN(num input, num n) {
    if (input < 0 && n > 0) {
      return false;
    } else if ((input == 0) | (n == 0)) {
      return false;
    } else if (input > 0 && n > 0) {
      return isMathematicalInteger(logBase(n, input));
    } else if (input < 0 && n < 0) {
      return logBase(-n, -input) % 2 == 1;
    } else {
      return logBase(-n, input) % 2 == 0;
    }
  }

  ///Returns log given ```base``` and ```input```.
  double logBase(num base, num input) {
    return math.log(input) / math.log(base);
  }

  ///Returns whether given double is a mathematical integer (e.x. 6.0).
  bool isMathematicalInteger(num input) {
    return input % 1 == 0.0;
  }

  ///Returns factorial given integer ```input```.
  int factorial(int input) {
    if (input < 0) {
      throw ArgumentError(
          'factorial($input) is undefined for negative arguments.');
    }
    if (input < factorials.length) {
      return factorials[input];
    }
    var r = factorials.last;
    for (var i = factorials.length; i <= input; i++) {
      r *= i;
      factorials.add(r);
    }
    return r;
  }

  /// A list of provided factorials, where ```factorials[i] == i!```
  ///
  /// This list has every factorial less than 2^64 (20! and under). The
  /// `factorial` method wraps this in some safety; it throws on negative
  /// numbers with an explanation, and multiplies to attempt to calculate
  /// factorials higher than 20!. The Dart language does not currently natively
  /// support 64+ bit integers, but if that happens in the future, the factorial
  /// method should still work.
  static List<int> factorials = [
    1,
    1,
    2,
    6,
    24,
    120,
    720,
    5040,
    40320,
    362880,
    3628800,
    39916800,
    479001600,
    6227020800,
    87178291200,
    1307674368000,
    20922789888000,
    355687428096000,
    6402373705728000,
    121645100408832000,
    2432902008176640000,
  ];

  ///Returns true if ```x``` is within some tolerance of ```y```
  ///
  ///Ex. 1: fuzzyEquals(1, 4, 3) returns true
  ///
  ///Ex. 2: fuzzyEquals(1, 6, 3) returns false
  ///
  ///Tolerance must be positive
  bool fuzzyEquals(num x, num y, num tolerance) {
    if (tolerance < 0) {
      throw ArgumentError('Tolerance must be >= 0');
    }
    var xL = x + tolerance;
    var xS = x - tolerance;
    return xL >= y && y >= xS;
  }

  ///Tests if the number `n` is (probably) a prime.
  ///
  ///This variant of the probabilistic prime test by Miller–Rabin is deterministic.
  ///
  ///It has been verified to return correct results for all `n` < 341,550,071,728,321.
  bool isPrime(int n) {
    if (n == 2 || n == 3 || n == 5) {
      return true;
    }
    if (n < 2 || n % 2 == 0 || n % 3 == 0 || n % 5 == 0) {
      return false;
    }
    if (n < 25) {
      return true;
    }
    var d = n - 1, s = 0;
    while (d % 2 == 0) {
      d ~/= 2;
      s++;
    }
    loop:
    for (final a in [2, 3, 5, 7, 11, 13, 17]) {
      var x = a.modPow(d, n);
      if (x == 1 || x == n - 1) {
        continue loop;
      }
      for (var r = 0; r <= s - 1; r++) {
        x = x.modPow(2, n);
        if (x == 1) {
          return false;
        }
        if (x == n - 1) {
          continue loop;
        }
      }
      return false;
    }
    return true;
  }

  ///Returns least common multiple of inputs.
  int lcm(int x, int y) => x * y ~/ x.gcd(y);

  ///Computes hyperbolic arc-cosine of a number.
  double acosh(num x) => math.log(x + math.sqrt(x * x - 1));

  ///Computes hyperbolic arc-sine of a number.
  double asinh(num x) => x.isInfinite && x.isNegative
      ? x.toDouble()
      : math.log(x + math.sqrt(x * x + 1));

  ///Computes hyperbolic arc-tangent of a number.
  double atanh(num x) => math.log((1 + x) / (1 - x)) / 2;

  ///Computes hyperbolic cosine of a number.
  double cosh(num x) {
    final y = math.exp(x);
    return (y + 1 / y) / 2;
  }

  ///Computes hyperbolic sine of a number.
  double sinh(num x) {
    final y = math.exp(x);
    return (y - 1 / y) / 2;
  }

  ///Computes hyperbolic tangent of a number.
  double tanh(num x) {
    final a = math.exp(x), b = math.exp(-x);
    return a.isInfinite
        ? 1
        : b.isInfinite
            ? -1
            : (a - b) / (a + b);
  }

  /// Best method for computing the arithmetic mean.
  ///
  /// The alternative ```(x + y) / 2``` fails for large values, which this handles.
  int mean(int x, int y) {
    return (x & y) + ((x ^ y) >> 1);
  }

  ///Combinations function from probability, nCr.
  ///
  ///Also described as the binomial coefficient.
  int combinationsOf(int n, int r) {
    return (factorial(n) / (factorial(n - r) * factorial(r))).round();
  }

  ///Permutations function from probability, nPr.
  int permutationsOf(int n, int r) {
    return (factorial(n) / factorial(n - r)).round();
  }

  /// Compute the signum of a number.
  /// The signum is -1 for negative numbers, +1 for positive numbers and 0 otherwise.
  num signum(num a) {
    return (a < 0.0) ? -1.0 : ((a > 0.0) ? 1.0 : a);
  }
}


lib/src/math/collectionutils.dart
/*
Starfruit, a set of Dart utility libraries.
Copyright (C) 2020 Aditya Kishore

This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
*/

import 'dart:math';

///Default implementation of StarCollectionsUtils().
_StarCollectionsUtils get collectionUtils => _StarCollectionsUtils();

///A class of packaged collection utilities and various commonplace functions not
///provided by (or improving upon) Dart functions. They can be used the default
///construction of StarCollectionsUtils, cUtils.
class _StarCollectionsUtils {
  ///Separate ```inp``` into a ```List``` of multiple, ```chunkSize``` ```List<Object>```'s.
  List<List<Object>> chunks (List<Object> inp, int chunkSize) {
    var i =0;
    var ret = <List<Object>>[];
    for (i = 0; i<inp.length; i+= chunkSize) {
      ret.add(inp.sublist(i, i+chunkSize));
      inp.removeRange(i, i+chunkSize);
    }
    ret.add(inp);
    return ret;
  }

  ///Collapse a ```List<List<Object>>``` into a ```List<Object>```.
  List<Object> collapse (List<List<Object>> inp) {
    var ret = <Object>[];
    for (var i in inp) {
      for (var x in i) {
        ret.add(x);
      }
    }
    return ret;
  }

  ///Return a list of every n'th (n = ```interval```) object from ```inp```.
  List<Object> nth (List<Object> inp, int interval) {
    var ret = <Object>[];
    for (var i = interval-1; i< inp.length; i+= interval) {
      ret.add(inp[i]);
    }
    return ret;
  }

  ///Return ```n``` random elements from ```inp```
  List<Object> random (List<Object> inp, [int n = 1]) {
    var ret = <Object>[];
    var random = Random.secure();
    for (var i = 0; i<n; i++) {
      ret.add(inp[random.nextInt(inp.length)]);
    }
    return ret;
  }

  ///Takes a ```Map``` and flips the keys and values.
  Map<Object,Object> flip (Map<Object, Object> inp) {
    var x = inp.keys.toList();
    var y = inp.values.toList();
    var ret = zip(y,x);
    return ret;
  }

  ///Zips two ```Lists``` into a ```Map```.
  Map<Object, Object> zip (List<Object> keys, List<Object> values) {
    var ret = <Object,Object>{};
    for (var i = 0; i<keys.length; i++) {
      var temp = <Object,Object>{keys[i]: values[i]};
      ret.addAll(temp);
    }
    return ret;
  }

  ///Unzips a ```Map``` into two ```Lists``` and returns a ```List``` containing the two ```Lists```.
  List<List<Object>> unzip (Map<Object,Object> inp) {
    var ret = <List<Object>>[];
    ret.add(inp.keys.toList());
    ret.add(inp.values.toList());
    return ret;
  }

  ///Checks if every ```Object``` in ```inp``` satisfies test
  bool every (List<Object> inp, bool Function(Object element) test) {
    for (var i in inp) {
      if (!test(i)) {
        return false;
      }
    }
    return true;
  }
}

lib/src/math/stats/statsutils.dart
/*
Starfruit, a set of Dart utility libraries.
Copyright (C) 2020 Aditya Kishore

This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
*/

import 'dart:math';

import '../mathutils.dart';

///A class that takes in a set of data as its constructor and can perform various
///statistical utilities with it.
extension StarStats on List<num> {
  ///Returns the median of this list.
  num get median {
    sort();
    if (length.isOdd) {
      return this[(length/2).floor()];
    }
    else {
      var x = this[(length/2).floor()];
      var y = this[(length/2).floor()-1];
      if (x is int && y is int) return mathUtils.mean(x, y);
      return (x + y) / 2;
    }
  }

  ///Returns the mean of this list.
  num get mean {
    num sum = 0;

    for (var i in this) {
      sum += i;
    }

    return sum / length;
  }

  /// Returns the mode of this list.
  ///
  /// If there are multiple modes, this returns the first mode.
  num get mode {
    num maxValue = 0;
    int maxCount = 0;

    for (var i = 0; i < length; ++i) {
      var count = 0;
      for (var j = 0; j < length; ++j) {
        if (this[j] == this[i]) {
          ++count;
        }
      }

      if (count > maxCount) {
        maxCount = count;
        maxValue = this[i];
      }
    }

    return maxValue;
  }

  /// Returns the variance of this list, assuming this list is a sample.
  ///
  /// See [popVariance] for the assumption that this list is a population.
  num get variance {
    num totvar = 0;
    var avg = mean;
    for (var i in this) {
      totvar += pow(i-avg, 2);
    }
    return totvar/(length-1);
  }

  /// Returns the variance of this list, assuming this list is a population.
  ///
  /// See [variance] for the assumption that this list is a sample.
  num get popVariance {
    num totvar = 0;
    var avg = mean;
    for (var i in this) {
      totvar += pow(i-avg, 2);
    }
    return totvar/length;
  }

  /// Returns the standard deviation of given set, assuming this list is a sample.
  ///
  /// See [popStdDev] for the assumption this is a population.
  num get stdDev {
    return sqrt(variance);
  }

  ///Returns the standard deviation of given set, assuming this list is a population.
  ///
  /// See [stdDev] for the assumption this is a sample.
  num get popStdDev {
    return sqrt(variance * ((length - 1) / length));
  }

  ///Returns the number of unique numbers in given set (a.k.a. cardinality).
  num get cardinality {
    var covdelems = <num>{};
    for (var i in this) {
      covdelems.add(i);
    }
    return covdelems.length;
  }

  ///Returns the k greatest elements of given set.
  List<num> topElements (int k) {
    var copy = toList(growable: false);
    copy.sort();
    return copy.sublist(length-k);
  }

  ///Returns the k least elements of given set.
  List<num> bottomElements (int k) {
    var copy = toList(growable: false);
    copy.sort();
    return copy.sublist(0, k);
  }
}

example/starfruit_example.dart
import 'package:starfruit/starfruit.dart';
import 'dart:core';

void main () {
  //Begin math functions

  //Round to 2 decimal places
  //Ceiling and flooring to decimal places is also available
  print('Round 3.5634 to 2 decimal places:');
  print(mathUtils.roundToDouble(1.005, 2));
  print('');

  //Power of two check
  print('Is 8 a power of two?');
  print(mathUtils.isPowerOfTwo(8));
  print('');

  //Calculate log base 2 of 8
  print('Calculate log base 2 of 8:');
  print(mathUtils.logBase(2, 8));
  print('');

  //double isInteger check
  print('Is 6.0 an integer?');
  print(mathUtils.isMathematicalInteger(6.0));
  print('');

  //Calculate 5!
  print('Calculate 5 factorial:');
  print(mathUtils.factorial(5));
  print('');

  //Check if 2 is within 3 of 5
  print('Is 2 within 3 of 5?');
  print(mathUtils.fuzzyEquals(2, 5, 3));
  print('');

  //Check if 7919 is a prime number
  print('Is 7919 prime?');
  print(mathUtils.isPrime(7919));
  print('');

  //LCM of 2 and 7
  print('Calculate LCM of 2 and 7:');
  print(mathUtils.lcm(2, 7));
  print('');

  //sinh of 1 radian
  print('Calculate sinh of 1 radian:');
  print(mathUtils.sinh(1));
  print('');

  //mean of 3 and 7
  print('Calculate mean of 3 and 7:');
  print(mathUtils.mean(3,7));
  print('');

  //nCr where n = 5 and r = 2
  print('Calculate the binomial coefficient of 2 and 5:');
  print(mathUtils.combinationsOf(5, 2));
  print('');

  //nPr where n = 5 and r = 2
  print('Calculate the permutations of 2 and 5:');
  print(mathUtils.permutationsOf(5, 2));
  print('');

  //end math functions

  //start collection functions

  //Separate List into chunks
  print('Separate List into chunks of 4:');
  print(collectionUtils.chunks([1,2,3,4,5,6,7,8], 4));
  print('');

  //Collapse List<List<Object>> into just List<Object>
  print('Collapse List<List> into List:');
  print(collectionUtils.collapse( [[1,2,3],[4,5,6,7,8]] ));
  print('');

  //Get every nth object of list
  print('Get every 4th object:');
  print(collectionUtils.nth([1,2,3,4,5,6,7,8], 4));
  print('');

  //Get every n random elements of list
  print('Get 5 random elements:');
  print(collectionUtils.random([1,2,3,4,5,6,7,8], 5));
  print('');

  //Flip keys and values of a map
  print('Flip keys and values of a map:');
  print(collectionUtils.flip({1:0, 2:8, 3:4, 0:5}));
  print('');

  //Zip two Lists into a map
  print('Zip two lists into a map:');
  print(collectionUtils.zip([1,2,3,4], [5,6,7,8]));
  print('');

  //Unzip a map into two lists
  print('Unzip a map into two lists:');
  print(collectionUtils.unzip({0:1, 8:2, 4:3, 5:0}));
  print('');

  //end collection functions

  //start stats functions
  var stats = [1,3,4,5,12,3,4,67,8,0,22];

  //median
  print('Calculate the median:');
  print(stats.median);
  print('');

  //mean
  print('Calculate the mean:');
  print(stats.mean);
  print('');

  //mode
  print('Calculate the mode:');
  print(stats.mode);
  print('');

  //variance
  print('Calculate the variance:');
  print(stats.variance);
  print('');

  //standard deviation
  print('Calculate the standard deviation:');
  print(stats.stdDev);
  print('');

  //cardinality; number of unique elements
  print('Calculate the cardinality:');
  print(stats.cardinality);
  print('');

  //7 greatest elements
  print('Return the 7 greatest elements:');
  print(stats.topElements(7));
  print('');

  //end stats functions

  //start xy stats functions

  var xystats = {2:3, 7:4, 4:6, 8:9, 1:2, 3:5, 11:14, 12:18};

  //Get correlation coefficient
  print('Calculate correlation coefficient:');
  print(xystats.corCoefficient);
  print('');

  //Get determination coefficent
  print('Calculate determination coefficient:');
  print(xystats.detCoefficient);
  print('');

  //Get adjusted determination coefficient
  print('Calculate adj. determination coefficient:');
  print(xystats.adjDetCoefficient);
  print('');

  //Get linear regression line
  print('Calculate linear regression line in form y = mx + b:');
  var lr = xystats.linReg;
  print('y = ${lr[0]}x + ${lr[1]}');
  print('');

  //Get quadratic regression line
  print('Calculate quadratic regression line in form y = ax^2 + bx + c:');
  var qdr = xystats.quadReg;
  print('y = ${qdr[0]}x^2 + ${qdr[1]}x + ${qdr[2]}');
  print('');

  //Get cubic regression line
  print('Calculate cubic regression line in form y = ax^3 + bx^2 + cx + d:');
  var cr = xystats.cubicReg;
  print('y = ${cr[0]}x^3 + ${cr[1]}x^2 + ${cr[2]}x + ${cr[3]}');
  print('');

  //Get quartic regression line
  print('Calculate quartic regression line in form y = ax^4 + bx^3 + cx^2 + dx + e:');
  var qrr = xystats.quarReg;
  print('y = ${qrr[0]}x^4 + ${qrr[1]}x^3 + ${qrr[2]}x^2 + ${qrr[3]}x + ${qrr[4]}');
  print('');
  // end xy stats functions
}